# 項目：評估和清理英國電商公司銷售數據

## 分析目標

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 簡介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 讀取數據

In [2]:
import pandas as pd

In [3]:
original_data = pd.read_csv("e_commerce.csv")

In [4]:
original_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


## 評估數據

In [5]:
original_data.sample(15)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
496151,578344,22493,PAINT YOUR OWN CANVAS SET,1,11/24/2011 9:21,3.29,NaN,United Kingdom
188668,553054,21866,UNION JACK FLAG LUGGAGE TAG,2,5/13/2011 10:22,1.25,17886.0,United Kingdom
42846,540023,22771,CLEAR DRAWER KNOB ACRYLIC EDWARDIAN,12,1/4/2011 12:58,1.25,15039.0,United Kingdom
177755,552180,21411,GINGHAM HEART DOORSTOP RED,3,5/6/2011 13:35,4.25,18235.0,United Kingdom
320,536401,22100,SKULLS SQUARE TISSUE BOX,2,12/1/2010 11:21,1.25,15862.0,United Kingdom
43726,540120,22245,"HOOK, 1 HANGER ,MAGIC GARDEN",48,1/5/2011 10:08,0.85,17675.0,United Kingdom
133179,547729,21707,FOLDING UMBRELLA BLACKBLUE POLKADOT,2,3/25/2011 10:52,4.95,15529.0,United Kingdom
43260,540026,22214,CANDLE PLATE LACE WHITE,5,1/4/2011 13:25,1.66,NaN,United Kingdom
228719,556942,84879,ASSORTED COLOUR BIRD ORNAMENT,16,6/15/2011 16:42,1.69,14222.0,United Kingdom
188220,553017,23209,LUNCH BAG DOILEY PATTERN,3,5/12/2011 19:01,1.65,12757.0,Portugal


从抽样的15行数据数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，
每列是交易相关的各个变量，因此不存在结构性问题。

<font size="5">評估數據乾淨度</font>

In [7]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值，而Description、CustomerID变量存在缺失值。

此外，InvoiceDate的数据类型应为日期，CustomerID的数据类型应为字符串，应当进行数据格式转换。

<font size = "5">評估缺失數據</font>

In [8]:
original_data[original_data['Description'].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/2011 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/2011 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/2011 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom


查看是否有Description有缺失但Unit price不為0的數據

In [10]:
original_data[(original_data['Description'].isnull()) & (original_data['UnitPrice'] != 0) ]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


Description有缺失但Unit price都為0,且對於我們分析暢銷產品並無用處，後續可刪除

CustomerID变量同样存在缺失值，因此也根據條件提取出缺失觀察值。

In [11]:
original_data[(original_data['Description'].isnull()) & (original_data['CustomerID'].isnull()) ]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/2011 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/2011 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/2011 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom


CustomerID的缺失直不影響我們對於暢銷產品的分析，故保留

<font size = "5">評估重複數據</font>

In [ ]:
根据数据变量的含义来看，虽然InvoiceNo、StockCode和CustomerID都是唯一标识符，但一次交易可能包含多件商品，因此InvoiceNo可以存在重复，
不同交易可以包含同件商品，因此StockCode可以存在重复，顾客可以进行多次交易或下单多个商品，因此CustomerID也可以存在重复。那么针对此数据集，
我们无需评估重复数据。

<font size = "5">評估不一致數據</font>

In [ ]:
不一致數據可能存在于Country变量中，我们要查看是否存在多个不同值指代同一國家的情况。

In [12]:
original_data['Country'].value_counts()

United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68
European

從Country变量值来看，"USA"、"United States"均在表示美国，"United Kingdom"、"UK"、"U.K."均在表示英国，
因此應該在清洗步驟對這些值進行统一，只保留一个指代值。

<font size = "5">評估錯誤數據</font>

In [4]:
original_data.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


從輸出結果来看，Quantity和UnitPrice存在負數，會對後續數值分析產生影響。

因此，我们先篩選出Quantity數值為負數的觀察值，進一步評估其含義。

In [5]:
original_data[original_data['Quantity']<0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


In [8]:
original_data[(original_data['Quantity']<0) & (original_data['InvoiceNo'].str[0]!='C') ]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,12/1/2010 16:50,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,12/2/2010 14:42,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,12/3/2010 15:30,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535333,581210,23395,check,-26,12/7/2011 18:36,0.0,NaN,United Kingdom
535335,581212,22578,lost,-1050,12/7/2011 18:38,0.0,NaN,United Kingdom
535336,581213,22576,check,-30,12/7/2011 18:38,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,12/8/2011 9:56,0.0,NaN,United Kingdom


以上猜想被证实错误，因为还存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

但根据以上输出结果，这些筛选出的观察值的UnitPrice观察值均为0，因此增加UnitPrice的条件进行验证。

In [9]:
original_data[(original_data['Quantity']<0) & (original_data['InvoiceNo'].str[0]!='C') & (original_data['UnitPrice'] != 0) ]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


根据输出结果，猜想得到验证，當Quantity变量为负数的时候，觀察值滿足以下條件之一：

InvoiceNo以C开頭，表訂單被取消
UnitPrice為0，說明單價為0英鎊
这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将Quantity为负数的观察值删除。

接下来筛选出UnitPrice为负数的观察值。

In [11]:
original_data[original_data['UnitPrice'] < 0 ]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,8/12/2011 14:51,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,8/12/2011 14:52,-11062.06,NaN,United Kingdom


In [ ]:
从输出结果来看，UnitPrice为负数的观察值都是坏账调账，不属于实际商品交易数据，因此也在数据清理步骤中也将其删除。

## 清理數據

根据前面評估部分得到的結論，我們需要進行的數据清理包括：

把InvoiceDate变量的数据类型转换为为日期时间
把CustomerID变量的数据类型转换为字符串
把Description变量缺失的观察值删除
把Country变量值"USA"替换为"United States"
把Country变量值"UK"、"U.K."替换为"United Kingdom"
把Quantity变量值为负数的观察值删除
把UnitPrice变量值为负数的观察值删除

为了區分开经过清理的数据和原始的数据，我们创建新的变量cleaned_data，让它为original_data复制出的副本。我们之后的清理步骤都将被运用在cleaned_data上。

In [13]:
cleaned_data = original_data.copy()
cleaned_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


把InvoiceDate变量的數据类型轉換為日期時間類型：

In [15]:
cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data['InvoiceDate'])
cleaned_data['InvoiceDate']

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把CustomerID变量的数据类型转换为字符串

In [16]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].astype(str)
cleaned_data['CustomerID'] 

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把CustomerID变量值結尾的".0"删除：

In [17]:
cleaned_data['CustomerID'].str.slice(0,-2)

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把Description变量缺失的观察值删除，并查看删除后该列空缺值个数和：

In [20]:
cleaned_data.dropna(subset=['Description'], inplace= True )

In [21]:
cleaned_data['Description'].isnull().sum()

0

把Country变量值"USA"替换为"United States"

In [22]:
cleaned_data['Country'] = cleaned_data['Country'].replace({'USA':'United States'})

In [27]:
len(cleaned_data[cleaned_data['Country']== 'USA'])

0

把Country变量值"UK"、"U.K."替换为"United Kingdom"

In [29]:
cleaned_data['Country'] = cleaned_data['Country'].replace({'UK':'United Kingdom','U.K.':'United Kingdom'})

In [30]:
print(len(cleaned_data[cleaned_data['Country']== 'U.K.']))
print(len(cleaned_data[cleaned_data['Country']== 'UK']))

0
0


把Quantity变量值為負數的觀察值删除(這邊反過來想，把Quantity>=0的值保留，即把<0的值刪除)

In [33]:
cleaned_data = cleaned_data[cleaned_data['Quantity'] >= 0 ]

In [35]:
len(cleaned_data[cleaned_data['Quantity'] < 0 ])

0

把UnitPrice变量值為負數的觀察值删除，檢查替換後UnitPrice變量為0的個數:

In [36]:
cleaned_data = cleaned_data[cleaned_data['UnitPrice'] >= 0 ]

In [38]:
len(cleaned_data[cleaned_data['UnitPrice'] < 0 ])

0

## 保存清理後的數據

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为e_commerce_cleaned.csv。

In [39]:
cleaned_data.to_csv('e_commerce_cleaned.csv', index = False)

In [40]:
pd.read_csv('e_commerce_cleaned.csv').head()

C:\Users\USER\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
